### DISCUSSION TITLE
Branching - Hosting provider integrations

### DISCUSSION THREAD

**Discussion by MildTomato** [supabase/supabase#18938] [Feature Requests]
This discussion is for tracking requests/demand for hosting providers that Supabase could integrate with so that Preview Branches can support their preview deployments.

Vote for Git providers [here](https://github.com/orgs/supabase/discussions/18936) and for general branching feedback, use [this discussion](https://github.com/orgs/supabase/discussions/18937).

**Comment by MildTomato** [comment]
Vercel

status: In active development

**Reply by tristdrum** [reply]
Very exciting potential here. Could be a beautiful pairing!

**Reply by gruckion** [reply]
Would be nice to get this.

**Reply by MildTomato** [reply]
Hey @tristdrum @gruckion 

This is in active development, we've tested it internally.
Should be supported in a week.

**Reply by gdaybrice** [reply]
@MildTomato hey, how does the migration align with the deploy to Vercel? Do they occur at the same time? Are they coordinated in any way to make sure the deploy happens after the migration (or vice-versa)?

Especially wondering how it behave in production vs preview as I imagine there might be a delay spawning a new preview environment.

**Reply by MildTomato** [reply]
There is a race condition with deployments right now. Vercel suggested that we use their webhooks to watch for deployments, and when a new unique one comes in for a git branch, we force re-deploy it for you so it has the latest environment variables. But we haven't got that working yet.

We're trying to figure out a better way of doing this; for now, probably makes sense we document this better so everyone knows the first Vercel deployment might need to be redeployed.

**Reply by gdaybrice** [reply]
Thanks, another issue is that the supabase integration sets the production variables in preview. If you aren't careful, the first time it would pickup the wrong environment if the branch one are not set yet.

**Reply by MildTomato** [reply]
> Thanks, another issue is that the supabase integration sets the production variables in preview. If you aren't careful, the first time it would pickup the wrong environment if the branch one are not set yet.

This is certainly not right - We're just looking at this now. I think we should fix the security aspect of this sharpish. 
I think the right outcome is we only set "production" env vars, and Preview Branches set "preview" env vars (based on a git branch). We shouldn't be setting Production based env vars for non "production" vercel targets.

If we change this, the 1st deploy will probably just fail to deploy as there will be a missing env var, but that's probably a better outcome for now than having production keys set for all env targets.

**Reply by MildTomato** [reply]
We have rolled a fix to stop setting all env targets in vercel for production based env vars.

For those currently using the Vercel Integration, you will need to manually untick the 'preview' and 'development' env targets in Vercel for the main production env vars if you don't want them used by 'preview' and 'development'.

What's changed:

- Any new connections made will insert the default production env vars to be set as only `'production'`. 
- It is possible that if Vercel deploys first before the env vars are inserted into the Vercel Project by Supabase, that first deployment will probably fail. You will need to manually redeploy that deployment in Vercel.

**Reply by gdaybrice** [reply]
If someone is in the same situation I've created a bash script to wait for the branch to be healthy, quite rudimentary but it works!

```
if [ "$#" -ne 1 ]; then
    echo "Usage: $0 branch_name"
    exit 1
fi

branch=$1
sleep_time=5

while true; do
    echo "Searching for branch: $branch"
    branch_id=$(pnpm supabase branches list --experimental | awk -v branch="$branch" 'BEGIN{FS="│"} $0 ~ branch {print $1}' | tr -d ' ' | tr -d '\n' | tr -d '\r' | sed 's/\x1b\[[0-9;]*m//g')

    if [ ! -z "$branch_id" ]; then
        echo "Found branch $branch with id: $branch_id"
        break
    fi

    sleep $sleep_time
done

while true; do
    echo "Checking status of branch: $branch"
    status=$(pnpm supabase branches get $branch_id --experimental | awk -F '│' 'NR>2 && !/STATUS/{print $7}' | tr -d ' ' | tr -d '\n' | tr -d '\r' | sed 's/\x1b\[[0-9;]*m//g')

    if [ "$status" = "ACTIVE_HEALTHY" ]; then
        echo "Branch $branch is active and healthy"
        break
    else
        echo "Current status of branch $branch: $status"
    fi

    sleep $sleep_time
done
```

**Reply by hyldmo** [reply]
For the vercel integration, would it be possible to expose a `SUPABASE_PROJECT_ID` env variable? I need to run codegen on the vercel host, and now i have to extract the project id from `SUPABASE_URL` which is a bit cumbersome. Alternatively, shouldn't `supabase gen types typescript --db-url $POSTGRES_URL` work when running on Vercel?

**Reply by leonard-henriquez** [reply]
@MildTomato would it be possible to join the alpha version ?

**Reply by aychang95** [reply]
@MildTomato Branching and integration with vercel is pretty sweet.

Seems that Supabase only creates a branching when a PR is created for the branch on github.
This causes our Vercel preview deployment to error out when a branch is pushed to remote because the preview env vars aren't available until after a PR is created.

Not sure if this is an isolated issue, intentional, or addressable via integration settings?

**Reply by la289** [reply]
How does the "branch linking" work. I.e. how does the supabase integration decide which env vars to assign to a given vercel branch? Not all hosted code changes will have db changes, and sometimes theres multiple vercel preview branches running at once.  What is the logic for the connection decision? is it manually set?

**Reply by kamilogorek** [reply]
> @MildTomato would it be possible to join the alpha version ?

@leonard-henriquez if you're on a paid plan, check your project 👀 

> This causes our Vercel preview deployment to error out when a branch is pushed to remote because the preview env vars aren't available until after a PR is created.

@aychang95 you should be able to set a fallbacks for `preview` env, and if there's no env variable set for the given branch, it will be picked up instead. At least afair this is how it used to work.

> How does the "branch linking" work. I.e. how does the supabase integration decide which env vars to assign to a given vercel branch?

@la289 it's all based on the GitHub connection right now. If you create a PR, and have both, GH and Vercel integrations, when a branch is created, it checks whether you have a Vercel connected as well, and if that's the case, it will update `preview` env vars with a git branch pulled from the PR assigned to it. So one branch preview = one vercel deployment.

**Reply by aychang95** [reply]
Thanks @kamilogorek!
So in order to not break commit-based Vercel branch deployments, a user should set fallback `preview` env vars?

That's doable, but I would have assumed that a git commit SHA creating a new branch instead of a PR would deploy a supabase branch. This sounds more in line with the term branching since most developers would require a new database for the branch they're working on.
The integration with Vercel could then trigger its own deployment with the new supabase env vars consistent with the branch the developer is working on. There could be a potential race condition with vercel and supabase, but I'm assuming the integration could handle the deployment in sequence?

**Reply by la289** [reply]
So this assumes you are deploying both vercel and supabase from the same repo right?

**Reply by kamilogorek** [reply]
> So in order to not break commit-based Vercel branch deployments, a user should set fallback preview env vars?

Correct.

> That's doable, but I would have assumed that a git commit SHA creating a new branch instead of a PR would deploy a supabase branch. This sounds more in line with the term branching since most developers would require a new database for the branch they're working on.

We only create branches for PRs, as this is the type of webhooks we're listening for. If we'd trigger branching for _every_ created branch, it could easily get out of hand and risk unwanted billing issues for customers.
You can still create branches manually for your branches without a PR, but you won't get automated Vercel syncing right now.

> So this assumes you are deploying both vercel and supabase from the same repo right?

No, that's not required. The flow is as follows:
- create GitHub PR
- have a github integration installed? great, lets create a branch preview
- branch preview is created
- do you have a vercel integration installed _on the same project that branch was created from_?
- if yes, sync vercel env variables with the branch name set to the one that was used for PR that triggered branch creation

The only required thing is that the same Supabase project is linked to Vercel and Github.

**Reply by hyldmo** [reply]
> We only create branches for PRs, as this is the type of webhooks we're listening for. If we'd trigger branching for _every_ created branch, it could easily get out of hand and risk unwanted billing issues for customers. You can still create branches manually for your branches without a PR, but you won't get automated Vercel syncing right now.

Any updates on this? I ended up creating branches via the CLI as I had some issues with the automatically created ones, but I would really like to get the variables for preview branches synced

**Reply by empz** [reply]
How does the sync of environment variables into Vercel work? Which variables needs to be overriden for preview environments? I successfully installed the Vercel integration but only Production variables are sync. I have a preview branch (which I created manually) that doesn't have any of the SUPABASE_ env. variables. Not sure how this is supposed to work.

**Reply by kamilogorek** [reply]
You can either mark a "preview" checkbox in the integration settings, which will repopulate the same env vars for each environment, or you can use it alongside GitHub integration, which will automatically set new branch env vars for every PR created.
We do not automatically sync env vars for manually created branches, as there has to be a Vercel deployment with the given name first, which is only created once Vercel detects new GitHub branch.

**Reply by empz** [reply]
Uhm, I created the branch manually and then created a PR and got the branch-specific environment variables populated into Vercel. So it seems the integration does sync vars even for manually created branches, but seems to be happening when the PR is created.

**Reply by kamilogorek** [reply]
Correct, the trigger used here is github webhook notifying us of the PR creation. We just dont sync manual branches at the time of their creation.

**Reply by johnford2002** [reply]
Cross posting a feedback post from the Branching thread since it seems more relevant to the Vercel integration piece.
https://github.com/orgs/supabase/discussions/18937#discussioncomment-10438263

The integration doesn't seem to clean up after itself when a PR is merged and the resulting branch is deleted from GitHub.

**Comment by MildTomato** [comment]
Netlify

status: Currently not in development

**Reply by ThatGuySam** [reply]
This is the big one for me.

Netlify has historically been very affordable for Open-Source projects and I end up using it commercial projects as well.

**Reply by MarkBennett** [reply]
Does Netlify expose the env values needed to access the branch in the api anywhere? If so, would you be open to a community implementation of this that could run in a GH Action?

We use Netlify and are looking at adding a db. Supabase is in the running but Deploy Previews are a huge part of our workflow so integration would be a big pro vs some of the alternatives. Currently comparing with AstroDB and Neon.

**Comment by MildTomato** [comment]
Render

status: Currently not in development

**Comment by MildTomato** [comment]
Cloudflare pages

status: Currently not in development

**Comment by remorses** [comment]
You are focusing on the wrong problems, let people change the env variables by themselves and make the GitHub app optional. Instead focus on making the core branching functionality work: add a merge button in the dashboard and collect all the bugs that happens when you try to merge, for example
- what happens if the schema migration fails?
- what happens if you delete a table with data in it? Should you add a error? Or confirmation message?
- how can you merge main back to a branch?


These are all things that PlanetScale implements, I suggest copying what they do and later add the GitHub app, only after you know the core model works

**Reply by ihadabs** [reply]
agree a branching like PlanetScale would be super nice

**Reply by sweatybridge** [reply]
We've started working on this to allow branching and merging migrations without GitHub. I will update here again when the feature is complete.

**Comment by LordZardeck** [comment]
A railway integration would be extremely helpful to us, or some API at least to fetch the env variables for a given branch name so we can make our own integration

**Reply by kamilogorek** [reply]
> or some API at least to fetch the env variables for a given branch name so we can make our own integration

https://api.supabase.com/api/v1#/environments/v1-list-all-branches + https://api.supabase.com/api/v1#/environments/v1-get-a-branch-config :)

**Comment by verbiiz** [comment]
I'm interested and super excited to be able to use this feature! would it be necessary to have multiple instances of environment variables stored in vercel? or would we be able to have the same environment variables across production, preview, and development (within vercel linked to supabase with supabase env vasrs) to still be able to use database branching?

also, how would database schematics definitions work? migration files are okay, but can they compare to, say, terraform/pulumi-style migration diffs for infra (or in this case, DB schema) changes? especially automatically and portably?

**Reply by verbiiz** [reply]
likewise, would be super nice to be able to have policies/row-level-security/etc described in code (and therefore version-controlled). at that point- templates would be so much more powerful as well, because you could define full suites of everything ready to go at the push of a button, and streamline having multiple dev environments for maximum data-security.

**Reply by kamilogorek** [reply]
> would it be necessary to have multiple instances of environment variables stored in vercel? or would we be able to have the same environment variables across production, preview, and development (within vercel linked to supabase with supabase env vasrs) to still be able to use database branching?

You can already choose what envs you want to sync to, and Vercel fallbacks to `preview` if there are no branch-specific env vars. However, we do not let you specify which env vars should be actually synced, so they'll always be creating a new instance for a new branch. They are always different though, so I'm not sure how exactly you'd like to share them.

> how would database schematics definitions work?

I'm not exactly sure what you're asking about here, but internally how we're using our own CLI to run everything on top newly created branch, so it works 1:1 how described in https://supabase.com/docs/guides/local-development/overview

> would be super nice to be able to have policies/row-level-security/etc described in code

Setting up RLS is nothing more than running its corresponding SQL queries, so you can have them stored and diffed inside migrations just fine. See https://supabase.com/docs/guides/database/postgres/row-level-security

**Reply by verbiiz** [reply]
@kamilogorek 

1. you can only have one set of supabase environment variables linked to vercel. now, what do you do when you want a separate database for prod and dev? 

2. yeah, i know what RLS is, however you have to go into the website to type all of the commands in instead of having it synchronized with your git repository. it's a lot of manual effort to handle diffs, but terraform among others have a version controlled infrastructure that can apply to this problem as well.

**Comment by k2xl** [comment]
Hi - I'm a bit confused on whether Vercel (assuming Vercel integration is installed) should be automatically getting environment variables from supabase when a branch is created (for that specific branch db)?

**Reply by k2xl** [reply]
Okay, I just figured out why it wasn't working for me. For anyone else seeing this. You have to open a PR. You can't just manually create the branches on both github and supabase. Let supabase trigger via a PR and it populates!

**Reply by kamilogorek** [reply]
That's exactly it. We're listening on GitHub `pull_request` webhook in order to trigger this behavior. Glad you found it.

**Reply by k2xl** [reply]
@kamilogorek Thanks - I do think that it should be clearer in the docs about this functionality and that explicitly creating the branch doesn't work.

Have you considered changing the behavior where it checks to see if the branch exists - and if so tries to populate the env vars (assuming they don't exist?)

**Reply by kamilogorek** [reply]
> - I do think that it should be clearer in the docs about this functionality and that explicitly creating the branch doesn't work.

https://supabase.com/docs/guides/deployment/branching#vercel

> Vercel GitHub integration also required.
> For branching to work with Vercel, you also need the [Vercel GitHub integration](https://vercel.com/docs/deployments/git/vercel-for-github).

It's there 😄

> Have you considered changing the behavior where it checks to see if the branch exists - and if so tries to populate the env vars (assuming they don't exist?)

I started working on expanding branching support to more providers and making it more generic, so this might naturally come to fruition in the future :)

**Reply by k2xl** [reply]
Thanks for the reply- the section you quote in the docs was not the part that was not clear.

I already had the integrations enabled.

1. In git I pushed a new staging branch. It appeared on vercel but did not auto create on supabase.

2. I tried manually creating the branch on supabase. Then i checked vercel and the env vars were still empty.

3. I then tried making a new change on the branch (thinking maybe because the branch didnt exist before and now it does) that it would. But it didnt.

4. I tried making the branch persistant. Resetting the branch. Trying combinations but still did not update env vars on vercel.

5. Finally i tried creating a pull request (i think i had deleted the branch at this point). NOW env vars were populated.

Steps 1-4 or some variation i imagine is what people would expect would work. Step 5 being the key could have been _more_ clear in the docs imo.

**Reply by kamilogorek** [reply]
Ah, gotcha. Updated the docs to reflect this https://github.com/supabase/supabase/pull/30969

**Reply by k2xl** [reply]
Thank you @kamilogorek ! much clearer!

**Comment by hbd** [comment]
Hey all, even after creating a PR, I'm not seeing env vars populate for my Vercel preview branch. I'm having to manually populate the env vars which requires resetting the Postgres password and constructing the connection string.

**Reply by kamilogorek** [reply]
Do you have both, GitHub and Vercel integration installed on the same project?

**Reply by k2xl** [reply]
@hbd If you made the same mistake as I did - you may need to "uncheck" Preview  from the env vars that you don't want enabled.

**Reply by hbd** [reply]
@kamilogorek, yes I do. The Preview deploys are going through when I open/push to a PR, however I'm seeing these issues:
* Env vars for preview branch not populating when PR is first opened
* Some values in Supabase are reset, like site URL, redirect URLs, and max storage upload limits, on the preview branch, every time a new commit is pushed to the feature branch PR

**Comment by chazwatkins** [comment]
Request: Fly.io Preview Apps

I didn't see this one in the previous comments.

**Comment by handlerda** [comment]
Is it safe to say that at this point this only works with Vercel?The docs are not super clear. Thank you. If not a blog / video showing how to enable this with hosting providers that support preview deployments would be helpful.

**Reply by kamilogorek** [reply]
No, it works with GitHub. Vercel just has a separate integration that encapsulates the behavior of branching for ease of use. You don't need it to have branches created, migrations running etc.
If you want to automate your deployment for any other provider right now, you can use our API directly https://api.supabase.com/api/v1#tag/environments

**Comment by KertLynx** [comment]
Request: Gitlab
I have been a Gitlab user for many years and I would be glad to see Supabase integrated with it.

**Reply by jgoux** [reply]
Hey @vscodesyncsettings, it's being actively worked on, aiming for Q1 this year! 👍

**Comment by adrianpunkt** [comment]
Railway! Similar to how Vercel works, Supabase could push it's config vars into the Railway PR deployment, that would be very much appreciated!
